doc/rxvtd.1.pod
=head1 NAME

@@RXVT_NAME@@d - @@RXVT_NAME@@ terminal daemon

=head1 SYNOPSIS

B<@@RXVT_NAME@@d> [-q|--quiet] [-o|--opendisplay] [-f|--fork] [-m|--mlock] [-e|--eval I<perlstring>]

B<@@RXVT_NAME@@d> -q -o -f    # for .xsession use

=head1 DESCRIPTION

This manpage describes the @@RXVT_NAME@@d daemon, which is the same vt102
terminal emulator as @@RXVT_NAME@@, but runs as a daemon that can open
multiple terminal windows within the same process.

You can run it from your X startup scripts, for example, although it is
not dependent on a working DISPLAY and, in fact, can open windows on
multiple X displays on the same time.

Advantages of running a @@RXVT_NAME@@ daemon include faster creation time
for terminal windows and a lot of saved memory.

The disadvantage is a possible impact on stability - if the
main program crashes, all processes in the terminal windows are
terminated. For example, as there is no way to cleanly react to abnormal
connection closes, C<xkill> and server resets/restarts will kill the
B<@@RXVT_NAME@@d> instance including all windows it has opened.

=head1 OPTIONS

B<@@RXVT_NAME@@d> currently understands a few options only. Bundling of
options is not yet supported.

=over 4

=item B<-q>, B<--quiet>

Normally, B<@@RXVT_NAME@@d> outputs the message C<< rxvt-unicode daemon
listening on <path> >> after binding to its control socket. This option
will suppress this message (errors and warnings will still be logged).

=item B<-o>, B<--opendisplay>

This forces B<@@RXVT_NAME@@d> to open a connection to the current
C<$DISPLAY> and keep it open.

This is useful if you want to bind an instance of B<@@RXVT_NAME@@d> to
the lifetime of a specific display/server. If the server does a reset,
B<@@RXVT_NAME@@d> will be killed automatically.

=item B<-f>, B<--fork>

This makes B<@@RXVT_NAME@@d> fork after it has bound itself to its control
socket.

=item B<-m>, B<--mlock>

This makes B<@@RXVT_NAME@@d> call mlockall(2) on itself. This locks
B<@@RXVT_NAME@@d> in RAM and prevents it from being swapped out to disk,
at the cost of consuming a lot more memory on most operating systems.

Note: In order to use this feature, your system administrator must have set
your user's RLIMIT_MEMLOCK to a size greater than or equal to the size of the
B<@@RXVT_NAME@@d> binary (or to unlimited). See F</etc/security/limits.conf>.

Note 2: There is a known bug in glibc (possibly fixed in 2.8 and later
versions) where calloc returns non-zeroed memory when mlockall is in
effect. If you experience crashes or other odd behaviour while using
--mlock, try it without it.

=item B<-e>, B<--eval> I<perlstring>

Evaluate the given perl code after basic initialisation (requires perl
support to be enabled when compiling B<@@RXVT_NAME@@d>).

This can be used for example to configure the internal perl interpreter,
which is shared between all terminal instances, or create additional
listening sockets for additional protocols.

The code is currently executed I<before> creating the normal listening
sockets: this might change in future versions.

=item B<-a>, B<--activate>

Support systemd socket activation (requires systemd support to be enabled
when compiling B<@@RXVT_NAME@@d>).

This option causes urxvtd to check on startup whether systemd passed in a
listening socket. If such a socket is found, urxvtd will use that socket instead
of creating its own. Otherwise, it will fall back to creating a new socket
as normal.

Note that the filesystem path used by systemd for the socket should match
the value of I<RXVT_SOCKET> in the user environment (or the default socket
path if I<RXVT_SOCKET> is unset; see "ENVIRONMENT" below).  This is to
ensure that B<@@RXVT_NAME@@c> can find the daemon. See I<doc/systemd>
for more information about how to configure urxvtd as a systemd service.

=back

=head1 EXAMPLES

This is a useful invocation of B<@@RXVT_NAME@@d> in a F<.xsession>-style
script:

   @@RXVT_NAME@@d -q -f -o

This waits till the control socket is available, opens the current display
and forks into the background. When you log-out, the server is reset and
B<@@RXVT_NAME@@d> is killed.

=head1 ENVIRONMENT

=over 4

=item B<RXVT_SOCKET>

Both B<@@RXVT_NAME@@c> and B<@@RXVT_NAME@@d> use the environment
variable F<RXVT_SOCKET> to create a listening socket and to contact
the @@RXVT_NAME@@d, respectively. If the variable is missing then
F<<< $HOME/.urxvt/urxvtd-I<< <nodename> >> >>> is used.

=item B<DISPLAY>

Only used when the C<--opendisplay> option is specified. Must contain a
valid X display name.

=back

=head1 SEE ALSO

@@RXVT_NAME@@(7), @@RXVT_NAME@@c(1)


src/rxvtd.C
/*----------------------------------------------------------------------*
 * File:	rxvtd.C
 *----------------------------------------------------------------------*
 *
 * All portions of code are copyright by their respective author/s.
 * Copyright (c) 2003-2007 Marc Lehmann <schmorp@schmorp.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------*/

#include "../config.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>

#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>

#if defined(ENABLE_FRILLS) && defined(_POSIX_MEMLOCK) && _POSIX_MEMLOCK > 0
# define ENABLE_MLOCK 1
#endif

#if ENABLE_MLOCK
# include <sys/mman.h>
#endif

#if ENABLE_SYSTEMD
#include <systemd/sd-daemon.h>
#endif

#include <errno.h>

#include "rxvt.h"
#include "rxvtdaemon.h"
#include "libptytty.h"

struct server : rxvt_connection {
  log_callback log_cb;
  getfd_callback getfd_cb;

  void read_cb (ev::io &w, int revents); ev::io read_ev;
  void log_msg (const char *msg);
  int getfd (int remote_fd);

  server (int fd)
  {
    read_ev.set <server, &server::read_cb> (this);
    log_cb.set  <server, &server::log_msg> (this);
    getfd_cb.set<server, &server::getfd>   (this);

    this->fd = fd;
    fcntl (fd, F_SETFD, FD_CLOEXEC);
    fcntl (fd, F_SETFL, 0);
    read_ev.start (fd, ev::READ);
  }

  void err (const char *format = 0, ...);
};

struct unix_listener {
  int fd;

  void accept_cb (ev::io &w, int revents); ev::io accept_ev;

  // create listener on a new socket
  unix_listener (const char *sockname);

  // create listener on an existing socket
  unix_listener (int descriptor);

  // copy constructor
  unix_listener (const unix_listener &other);
};

unix_listener::unix_listener (const char *sockname)
{
  accept_ev.set<unix_listener, &unix_listener::accept_cb> (this);

  sockaddr_un sa;

  if (strlen (sockname) >= sizeof(sa.sun_path))
    {
      fputs ("socket name too long, aborting.\n", stderr);
      exit (EXIT_FAILURE);
    }

  if ((fd = socket (AF_UNIX, SOCK_STREAM, 0)) < 0)
    {
      perror ("unable to create listening socket");
      exit (EXIT_FAILURE);
    }

  fcntl (fd, F_SETFD, FD_CLOEXEC);
  fcntl (fd, F_SETFL, O_NONBLOCK);

  sa.sun_family = AF_UNIX;
  strcpy (sa.sun_path, sockname);

  unlink (sockname);

  mode_t omask = umask (0077);

  if (bind (fd, (sockaddr *)&sa, sizeof (sa)))
    {
      perror ("unable to bind listening socket");
      exit (EXIT_FAILURE);
    }

  umask (omask);

  if (listen (fd, 5))
    {
      perror ("unable to listen on socket");
      exit (EXIT_FAILURE);
    }

  accept_ev.start (fd, ev::READ);
}

unix_listener::unix_listener (int descriptor)
{
  accept_ev.set<unix_listener, &unix_listener::accept_cb> (this);

  fd = descriptor;

  fcntl (fd, F_SETFD, FD_CLOEXEC);
  fcntl (fd, F_SETFL, O_NONBLOCK);

  accept_ev.start (fd, ev::READ);
}

unix_listener::unix_listener (const unix_listener &other)
{
  fd = other.fd;
}

void unix_listener::accept_cb (ev::io &w, int revents)
{
  int fd2 = accept (fd, 0, 0);

  if (fd2 >= 0)
    new server (fd2);
}

int server::getfd (int remote_fd)
{
  send ("GETFD");
  send (remote_fd);
  return ptytty::recv_fd (fd);
}

void server::log_msg (const char *msg)
{
  send ("MSG"), send (msg);
}

void server::err (const char *format, ...)
{
  if (format)
    {
      char err[1024];

      va_list ap;
      va_start (ap, format);
      vsnprintf (err, 1024, format, ap);
      va_end (ap);

      log_msg (err);
    }

  send ("END"), send (0);
  close (fd);
  delete this;
}

void server::read_cb (ev::io &w, int revents)
{
  auto_str tok;

  if (recv (tok))
    {
      if (!strcmp (tok, "NEW"))
        {
          stringvec *argv = new stringvec;
          stringvec *envv = new stringvec;

          for (;;)
            {
              if (!recv (tok))
                return err ();

              if (!strcmp (tok, "END"))
                break;
              else if (!strcmp (tok, "ENV") && recv (tok))
                envv->push_back (strdup (tok));
              else if (!strcmp (tok, "ARG") && recv (tok))
                argv->push_back (strdup (tok));
              else
                return err ("protocol error: unexpected NEW token.\n");
            }

          {
            rxvt_term *term = new rxvt_term;

            term->log_hook = &log_cb;
            term->getfd_hook = &getfd_cb;

            bool success = true;

            try
              {
                term->init (argv, envv);
              }
            catch (const class rxvt_failure_exception &e)
              {
                success = false;
              }

            term->log_hook = 0;

            chdir ("/"); // init might change to different working directory

            if (!success)
              term->destroy ();

            send ("END"); send (success ? 1 : 0);
          }
        }
      else if (!strcmp (tok, "QUIT"))
        _exit (0);
      else
        return err ("protocol error: request '%s' unsupported.\n", (char *)tok);
    }
  else
    return err ();
}

int
main (int argc, char *argv[])
{
  ptytty::init ();

  static char opt_fork, opt_opendisplay, opt_quiet;
#if ENABLE_PERL
  static char *opt_eval;
#endif
#if ENABLE_MLOCK
  static char opt_lock;
#endif
#if ENABLE_SYSTEMD
  static char opt_activate;
#endif

  for (int i = 1; i < argc; i++)
    {
      if (!strcmp (argv [i], "-f") || !strcmp (argv [i], "--fork"))
        opt_fork = 1;
      else if (!strcmp (argv [i], "-o") || !strcmp (argv [i], "--opendisplay"))
        opt_opendisplay = 1;
      else if (!strcmp (argv [i], "-q") || !strcmp (argv [i], "--quiet"))
        opt_quiet = 1;
#if ENABLE_MLOCK
      else if (!strcmp (argv [i], "-m") || !strcmp (argv [i], "--mlock"))
        opt_lock = 1;
#endif
#if ENABLE_PERL
      else if (!strcmp (argv [i], "-e") || !strcmp (argv [i], "--eval"))
        opt_eval = argv [++i];
#endif
#if ENABLE_SYSTEMD
      else if (!strcmp (argv [i], "-a") || !strcmp (argv [i], "--activate"))
        opt_activate = 1;
#endif
      else
        {
          rxvt_log ("%s: unknown option '%s', aborting.\n", argv [0], argv [i]);
          return EXIT_FAILURE;
        }
    }

  rxvt_init ();

#if ENABLE_PERL
  if (opt_eval)
    {
      rxvt_perl.init ();
      rxvt_perl.eval (opt_eval);
    }
#endif

  // optionally open display and never release it.
  if (opt_opendisplay)
    if (const char *dpy = getenv ("DISPLAY"))
      displays.get (dpy ? dpy : ":0"); // move string logic into rxvt_display maybe?

  char *sockname = rxvt_connection::unix_sockname ();
  int supplied_fd = 0;

#if ENABLE_SYSTEMD
  // optionally look for a socket file passed in by systemd.
  if (opt_activate)
    {
      int n = sd_listen_fds (1);
      if (n > 1)
        {
          fputs ("too many file descriptors received, aborting.\n", stderr);
          exit (EXIT_FAILURE);
        }
      else if (n == 1)
        supplied_fd = SD_LISTEN_FDS_START + 0;
    }
#endif

  unix_listener l = supplied_fd ?
    unix_listener (supplied_fd) :
    unix_listener (sockname);

  chdir ("/");

  if (!opt_quiet)
    {
      if (supplied_fd)
        puts ("rxvt-unicode daemon listening on received socket.");
      else
        printf ("rxvt-unicode daemon listening on %s.\n", sockname);
      fflush (stdout);
    }

  free (sockname);

  pid_t pid = 0;
  if (opt_fork)
    {
      pid = fork ();
    }

#if ENABLE_MLOCK
  // Optionally perform an mlockall so this process does not get swapped out.
  if (opt_lock && !pid)
    if (mlockall (MCL_CURRENT | MCL_FUTURE) < 0)
      perror ("unable to lock into ram");
#endif

  if (opt_fork)
    {
      if (pid < 0)
        {
          rxvt_log ("unable to fork daemon, aborting.\n");
          return EXIT_FAILURE;
        }
      else if (pid > 0)
        _exit (EXIT_SUCCESS);

      ev_loop_fork (EV_DEFAULT_UC);
    }

  ev_run ();

  return EXIT_SUCCESS;
}

